### ISSUE TITLE
The value (?) of an optimized bytecode interpreter

### ISSUE THREAD

**Issue by lukego** [raptorjit/raptorjit#172]
This issue is inspired by a discussion partly [on the LuaJIT mailing list](https://www.freelists.org/post/luajit/ANN-dumpanalyze-tool-for-working-with-LuaJIT-dumps,13).

Our experience with Snabb has been that bytecode interpreter performance is much too low to ever use in production. This means that we need to always make sure that all performance-sensitive code is successfully compiled. This also suggests that the bytecode interpreter should be engineered for primarily for maintainability since we don't get any practical benefit from performance optimizations.

@igelhaus reported a different experience. They see quite decent performance with the bytecode interpreter. They also sometimes deliberately fall back to interpretation in certain functions that are hard to optimize for the JIT.

So how can we reconcile these two very different experiences?

I have a hypothesis: that the bytecode interpreter performance is generally okay but that it is very low for code that uses the FFI.

Here is some supporting evidence:
1. Snabb code uses the FFI extensively and has awful interpreter performance, while @igelhaus code does not use the FFI and has decent interpreter performance.
2. Snabb code running in the interpreter is associated with very high rates of garbage collection overhead.
3. Operations on FFI pointers (e.g. assigning one to a local variable) in the interpreter will always force a heap allocation and garbage collection step ("boxing") but this is usually optimized away in JIT code ("allocation sinking.")

So it is possible that a major reason for Snabb seeing low interpreter performance is not the fundamental "mcode vs interp" overhead but rather the loss of allocation sinking optimization on FFI values.

This could have important implications for the RaptorJIT roadmap. One of the most important features that we want is unboxed FFI pointers i.e. to make FFI code run fast even when allocation sinking optimization doesn't kick in. We have talked about porting the VM to C to make this optimization easier to implement. However, if we don't have to box FFI pointers then maybe that means we will actually be able to use the bytecode interpreter! So then we would be interested in optimizing that and maybe that means porting it back to assembler again........

So overall it seems worth reconsidering the idea of keeping the VM written in assembler. Could be that this will actually be valuable in the future and worth the effort of maintaining while implementing new features. Have to think about this trade-off a bit more carefully.

**Comment by javierguerragiraldez** [comment]
i think (without any evidence whatsoever) that the high cost of FFI in the interpreter is not only because of allocations and constantly boxing/unboxing values, but also because any operation on FFI objects first have to decode the ctype information to find exactly what to do.

The extreme case might be calling a C function via FFI, it effectively does an `ffi.ew()` for each argument and the return value.  in the end it even calls the GC several times (one step for each argument)!

An obvious improvement for that would be to cache (or memoize) the ctype analyzing step, somehow recording what it ends up doing so the next time it only has to reexecute the same operation.  of course, because of dynamic values, it would first have to check that the ctype is really the same as before, and redo the analysis if not....

In short, trace the execution, emit optimized code for it, patch that into the original bytecode and insert a type guard to fallback if needed.   Why does it sound familiar?

**Comment by lukego** [comment]
@javierguerragiraldez Relatedly: @igelhaus said he likes to fallback to the interpreter in very branchy code to avoid excessive side-trace penalties. But maybe it's mostly a bug (#107) that taking side traces is expensive anyway and we should fix that instead?

**Comment by ldeniau** [comment]
I am not sure that I added the calculator to the right issue so let's link it here: #173 

Concerning the value of a fast VM, I think that if it is achievable in portable C with little constraints (see my example), I would definitely go for it. My experience with LuaJIT is not about the speed ratio between JITed and VM code, which is not so bad for simple micro benchmark, but more that VM does not optimise the code at all: all operations are blindly executed one after the other. Usually we assume as developper that the compiler will do some nice optimisations and keep the code clear. This is why the VM is inefficient for, just because the same optimisations are not applied. If the VM can perform some optimisations, and use higher level instructions and/or internal object representations, significant speed-up can be obtained (say within x1-x4 to compiled code) because most of the runtime will be spent in C or optimised instructions set. Of course this excludes optimisations based on runtime statistics or recording (otherwise let's the JIT do it). Moreover, some optimisations could be performed at the level of the parser if the code had an AST-like representation (partial and temporally).
